### PR TITLE
DBAL-1143 Added Postgres 9.4 platform

### PR DESCRIPTION
**Pull request by mbeccati**
Features and changes:
- jsonb can be used with: options={"jsonb"=true}
- OVER is no longer reserved
- LATERAL is now reserved

### CODE REVIEW

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @link http://www.postgresql.org/docs/9.4/static/sql-keywords-appendix.html
+     */
+    protected function getKeywords()
+    {
+        $parentKeywords = array_diff(parent::getKeywords(), array(
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php]: @deeky666 did we bump to PHP 5.4? Can we do that as well? (we cannot keep adding new platforms and having an increasingly growing matrix)

File: lib/Doctrine/DBAL/Platforms/PostgreSQL94Platform.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * @since  2.5
+ */
+class PostgreSQL94Platform extends PostgreSQL92Platform
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getJsonTypeDeclarationSQL(array $field)
+    {
+        if (!empty($field['jsonb'])) {
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL94Platform.php]: This new feature requires documentation

Comment (mbeccati) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL94Platform.php]: Should I do that? If so, any hint on how to do it? ;)

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+namespace Doctrine\DBAL\Platforms\Keywords;
+
+/**
+ * PostgreSQL 9.4 reserved keywords list.
+ *
+ * @author Matteo Beccati <matteo@beccati.com>
+ * @link   www.doctrine-project.org
+ * @since  2.5
```
Comment (deeky666) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php]: it has to be `2.6`

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php (hunk 1 above)
Comment (deeky666) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php]: @Ocramius no we haven't bumped to 5.4 yet. What do you mean by *we cannot keep adding new platforms and having an increasingly growing matrix*? And why is that related to PHP 5.4?

File: tests/Doctrine/Tests/DBAL/Platforms/PostgreSQL94PlatformTest.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function createPlatform()
+    {
+        return new PostgreSQL94Platform();
+    }
+
+    /**
+     * @group DBAL-553
```
Comment (deeky666) [inline tests/Doctrine/Tests/DBAL/Platforms/PostgreSQL94PlatformTest.php]: These group annotations are wrong. Please remove them.

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php (hunk 1 above)
Comment (zeroedin-bill) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php]: I think he's talking about the size of our test set -- We're testing on 41 combinations of platform and php version right now. Without 5.3 we'd be testing on 33. :neutral_face:

File: lib/Doctrine/DBAL/Platforms/PostgreSQL94Platform.php (hunk 2 above)
Comment (zeroedin-bill) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL94Platform.php]: You'll need to document the type here: https://github.com/doctrine/dbal/blob/master/docs/en/reference/types.rst

Comment (mbeccati) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL94Platform.php]: @zeroedin-bill Done, thanks for the suggestion.

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php (hunk 1 above)
Comment (Ocramius) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php]: @zeroedin-bill speaking of which: drop 5.3 and 5.4 for 2.6? [Nudge Nudge](https://www.youtube.com/watch?v=SrDFGa0juCM&t=96)

File: docs/en/reference/types.rst [hunk 5]
```diff
@@ -759,6 +763,10 @@ Please also notice the mapping specific footnotes for additional information.
 .. [18] Chosen if the column length is less or equal to **2 ^ 16 - 1 = 65535**.
 .. [19] Chosen if the column length is less or equal to **2 ^ 24 - 1 = 16777215**.
 .. [20] Chosen if the column length is less or equal to **2 ^ 32 - 1 = 4294967295** or empty.
+.. [21] Chosen if the column definition does not contain the **jsob** option inside the **platformOptions**
```
Comment (zeroedin-bill) [inline docs/en/reference/types.rst]: Typo: jsonb

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php (hunk 1 above)
Comment (zeroedin-bill) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL94Keywords.php]: 2.6 probably won't be out by the end of 5.4 security support, so I can't really object to the idea...

File: docs/en/reference/types.rst (hunk 5 above)
Comment (mbeccati) [inline docs/en/reference/types.rst]: Fixed.

**Comment by texdc** [comment]
ping. [9.5 alpha](http://www.postgresql.org/about/news/1595/) was just released today.

**Comment by mbeccati** [comment]
I totally forgot about this one! I should have properly harassed @Ocramius about it at the last phpDay ;)

@texdc that said, I don't see anything in 9.5 that would require changes in dbal, but please correct me if I'm wrong. Also there are no new reserved keywords, according to https://github.com/postgres/postgres/commits/master/doc/src/sgml/keywords.sgml

**Comment by JoshZA** [comment]
Please merge this, we need jsonb support :smiley:

**Comment by mbeccati** [comment]
FYI, I've squashed and rebased the PR.

**Comment by m0ppers** [comment]
bump. need jsonb as well :hourglass_flowing_sand:  :D

**Comment by Ocramius** [comment]
:+1:

**Pull request merged**